### PR TITLE
Max salary can't be more than twice the min salary

### PR DESCRIPTION
**Pull request by ignasi35**
Limiting the salary range so that Max is not more than twice Min sounds reasonable enough.

We must draw the line somewhere...

### CODE REVIEW

**Comment by ignasi35** [comment]
Linting on each PR FTW! 😄

**Review by oriolblanc — APPROVED** [review]
🚀👏

**Review by gonzaloserrano — APPROVED** [review]

**Comment by gonzaloserrano** [comment]
Let me tell the rest to review, for me is ✅

**Review by sdecandelario — APPROVED** [review]

**Comment by linkita** [comment]
I disagree. This means that a range 30-60 is not accepted by candebot. We have very few cases with very big salary ranges.

**Comment by ignasi35** [comment]
30-60 _should_ be valid (unless the code in this PR is wrong).

I browsed the history of job offers and only the latest (salary range 20-100) would have been forbidden. I even chose the linit at 2x because I noticed a couple like 50-100 or 60-120.

I am happy to relax the limit a bit further to 2.5x or even 3x but I think job offers with such big salary ranges are not very informative.

**Comment by ignasi35** [comment]
Hmm, I just realised this change makes MinSalary effectively compulsory (I think it's optional at the moment).

I will have to consider the case where MinSalary is blank.

**Comment by gonzaloserrano** [comment]
I still think we need something to avoid crazy ranges, so I did this table so we can choose the multiplier:
CleanShot 2021-09-22 at 09 57 18

30-60 may be OK'ish (not sure), but 30-69 is OK? They look like total different roles to me.

File: bot/interact.go [hunk 1]
```diff
@@ -81,6 +81,10 @@ func interactAPIHandler(botContext cmd.BotContext) http.HandlerFunc {
 					if minSalary > maxSalary {
 						validationErrors["min_salary"] = "The Salary Min field should contain a lower value than the specified in Salary Max field."
 					}
+
+					if (minSalary * 2) > maxSalary {
+						validationErrors["max_salary"] = "The Salary Max field should be at most twice the value specified in Salary Min field."
```
Comment (ignasi35) [inline bot/interact.go:86]: Creo que en vez de un mensaje de error con la regla de negocio, un mensaje de error con una explicación puede ser más valioso:

```suggestion
			 			validationErrors["max_salary"] = "The gap between MinSalary and MaxSalary is rather large. Maybe you should post two different job offers with different responsibilities and required qualifications. Salary is a relevant field, we recommend you try to keep it meaningful to increase the chances of taking the position seriously by potential candidates."
```

**Comment by smoya** [comment]
Remember whatever rule we choose here, should be explained to the user who post. I mean, in a very concise an easy wording that can be understand by any kind of public.
In other words, no mathematicians are going to submit offers.

**Comment by sdecandelario** [comment]
For me it's ok to go with 2.5 I think can give more range and be less restrictive

File: bot/interact.go [hunk 2]
```diff
@@ -81,6 +81,10 @@ func interactAPIHandler(botContext cmd.BotContext) http.HandlerFunc {
 					if minSalary > maxSalary {
 						validationErrors["min_salary"] = "The Salary Min field should contain a lower value than the specified in Salary Max field."
 					}
+
+					if (minSalary * 2) > maxSalary {
+						validationErrors["max_salary"] = "The gap between MinSalary and MaxSalary is rather large. Maybe you should post two different job offers with different responsibilities and required qualifications. Salary is a relevant field, we recommend you try to keep it meaningful to increase the chances of taking the position seriously by potential candidates."
```
Comment (gonzaloserrano) [inline bot/interact.go:86]:
```suggestion
					if (minSalary * 2.5) > maxSalary {
						validationErrors["max_salary"] = "The gap between MinSalary and MaxSalary is rather large. Maybe you should post two different job offers with different responsibilities and required qualifications. Salary is a relevant field, we recommend you try to keep it meaningful to increase the chances of taking the position seriously by potential candidates."
```

Comment (gonzaloserrano) [reply bot/interact.go:86]: cc @linkita @smoya @KoeSystems (rest agrees)

Comment (ignasi35) [reply bot/interact.go:86]: @gonzaloserrano my GoLang _fu_ is non-existant and this change had no test so my first draft to move from `2x` to `2.5x` was using:  `(minSalary * 5) > (maxSalary * 2)`.

I trust you know what you're doing. :-D

Comment (gonzaloserrano) [reply bot/interact.go:86]: > I trust you know what you're doing. :-D

I have no idea, just changed the value XD

Comment (gonzaloserrano) [reply bot/interact.go:86]: well, and also the type... I'll double check

Comment (gonzaloserrano) [reply bot/interact.go:86]: it does now work 

> interact.go|81 col 12| constant 2.5 truncated to integer 🤦

Comment (ignasi35) [reply bot/interact.go:86]: LOL
YOLO based programming, test in prod, hahaha

I'm not at my keyboard ATM. I'll Push my draft later.

**Comment by ignasi35** [comment]
this is ready for a final review

File: bot/interact.go [hunk 3]
```diff
@@ -81,6 +81,10 @@ func interactAPIHandler(botContext cmd.BotContext) http.HandlerFunc {
 					if minSalary > maxSalary {
 						validationErrors["min_salary"] = "The Salary Min field should contain a lower value than the specified in Salary Max field."
 					}
+
+					if (minSalary * 5) > (maxSalary * 2) {
```
Comment (gonzaloserrano) [inline bot/interact.go:85]: this should be more understandable
```suggestion
					if 2.5*float32(minSalary) > float32(maxSalary)  {
```

Comment (ignasi35) [reply bot/interact.go:85]: Whichever is more idiomatic. I'm fine with either option.

**Review by gonzaloserrano — APPROVED** [review]

File: bot/interact.go [hunk 4]
```diff
@@ -81,6 +81,10 @@ func interactAPIHandler(botContext cmd.BotContext) http.HandlerFunc {
 					if minSalary > maxSalary {
 						validationErrors["min_salary"] = "The Salary Min field should contain a lower value than the specified in Salary Max field."
 					}
+
+					if 2.5*float32(minSalary) > float32(maxSalary) {
```
Comment (ignasi35) [inline bot/interact.go:85]:
```suggestion
					if 2.5*float32(minSalary) < float32(maxSalary) {
```

Been doing some local testing. The condition was backwards! 😅

File: bot/interact.go [hunk 5]
```diff
@@ -81,6 +81,10 @@ func interactAPIHandler(botContext cmd.BotContext) http.HandlerFunc {
 					if minSalary > maxSalary {
 						validationErrors["min_salary"] = "The Salary Min field should contain a lower value than the specified in Salary Max field."
 					}
+
+					if 2.5*float32(minSalary) < float32(maxSalary) {
+						validationErrors["max_salary"] = "The gap between MinSalary and MaxSalary is rather large. Maybe you should post two different job offers with different responsibilities and required qualifications. Salary is a relevant field, we recommend you try to keep it meaningful to increase the chances of taking the position seriously by potential candidates."
```
Comment (smoya) [inline bot/interact.go:86]:
```suggestion
						validationErrors["max_salary"] = "The gap between MinSalary and MaxSalary is rather large. Maybe you should post two different job offers with different responsibilities and required qualifications. Salary is a relevant field; we recommend you keep it meaningful to increase the chances of taking the position seriously by potential candidates."
```

**Comment by smoya** [comment]
The reality is that I don't +1 this change since it introduces complexity. I'm afraid people will rather skip adding the min salary and that's all.
However, others did +1 so I can't do that much here.

**Comment by ignasi35** [comment]
(there's two different discussions)

> The reality is that I don't +1 this change since it introduces complexity. 

It does. It's a trade-off between keeping the job listing relevant VS keeping the code less-complex.

> The message recommends to split the offer but I'm afraid people will rather skip adding the min salary and that's all. 

I improvised when rewriting the message and I'm happy to remove the recommendation limiting the statement to the sentiment. I.e. remove `Maybe you should post two different job offers with different responsibilities and required qualifications.`

> However, others did +1 so I can't do that much here.

As the creator and most relevant contributor of the candebot I value your input and I'm happy to adapt the PR.

**Comment by gonzaloserrano** [comment]
> he message recommends to split the offer but I'm afraid people will rather skip adding the min salary and that's all.

I'd make the min salary mandatory too.

**Comment by ignasi35** [comment]
> I'd make the min salary mandatory too.

Topic for another discussion?

**Comment by gonzaloserrano** [comment]
The relevant part for this one is that I think salary_min optionality it's not a blocker.

**Comment by smoya** [comment]
> It does. It's a trade-off between keeping the job listing relevant VS keeping the code less-complex

Sorry I wasn't very clear! I didn't mean code complexity but operation-related.

> I.e. remove Maybe you should post two different job offers with different responsibilities and required qualifications.

I actually like the recommendation. I'm just playing the devil's advocate here and saying what it might happen in the end.

> I'd make the min salary mandatory too.

+1 to open another discussion.

> The relevant part for this one is that I think salary_min optionality it's not a blocker.

Totally true.

**Review by smoya — APPROVED** [review]
Let's try and see how it goes in the next days :)

Thanks for your work here @ignasi35

**Pull request merged**